plugins/nominal-connection-checker/src/type_hierarchy.js

```diff
@@ -138,51 +178,64 @@ export class TypeHierarchy {
    * Link: https://bit.ly/2SrCRs5 .
    *
+   * But the above has been slightly modified to work for both ancestors and
+   * descendants.
+   *
    * Operates in O(nm) where n is the number of nodes and m is the number of
    * edges.
+   *
+   * @param {!Map<string, !Map<string, Array<string>>>} nearestCommonMap The
+   *      map of nearest common types (either ancestors or descendants) that we
+   *     are initializing.
+   * @param {function(TypeDef):!Array<string>} relevantRelatives Returns the
+   *     relatives that are relevant to this procedure. In the case of
+   *     ancestors, returns supertypes, and in the case of descendants, returns
+   *     subtypes.
+   * @param {function(TypeDef, string):boolean} isNearest Returns true if the
+   *     type associated with the string name is the nearest common X of the
+   *     type associated with the string name and the given type def.
    * @private
    */
-  initNearestCommonParents_() {
+  initNearest_(nearestCommonMap, relevantRelatives, isNearest) {
     const unvisitedTypes = new Set(this.types_.keys());
     while (unvisitedTypes.size) {
       for (const typeName of unvisitedTypes) {
         const type = this.types_.get(typeName);
-        const unvisitedSupers = type.supers().filter(
-            unvisitedTypes.has, unvisitedTypes);
-        if (unvisitedSupers.length) {
+        const hasUnvisited = !!relevantRelatives(type).filter(
+            unvisitedTypes.has, unvisitedTypes).length;
+        if (hasUnvisited) {
           continue;
         }
         unvisitedTypes.delete(typeName);
 
         const map = new Map();
-        this.nearestCommonParents_.set(typeName, map);
+        nearestCommonMap.set(typeName, map);
         for (const [otherTypeName] of this.types_) {
-          let leastCommonAncestors = [];
-          if (type.hasDescendant(otherTypeName)) {
-            leastCommonAncestors.push(typeName);
+          let nearestCommon = [];
+          if (isNearest(type, otherTypeName)) {
+            nearestCommon.push(typeName);
           } else {
-            // Get all the least common ancestors this type's direct
-            // ancestors have with the otherType.
-            type.supers().forEach((superTypeName) => {
-              leastCommonAncestors.push(
-                  ...this.nearestCommonParents_.get(superTypeName)
+            // Get all the nearest common types this type's relevant relatives
+            // have with the otherType.
+            relevantRelatives(type).forEach((relTypeName) => {
+              nearestCommon.push(
+                  ...nearestCommonMap.get(relTypeName)
                       .get(otherTypeName));
             });
-            // Only include types that have no descendants in the array.
-            leastCommonAncestors = leastCommonAncestors.filter(
+            // Remove types that have a nearer relative in the array.
+            nearestCommon = nearestCommon.filter(
                 (typeName, i, array) => {
```

Comment:
The `type1` and `type2` is nice since it's shorter (and I don't think we're getting much by adding `nearestType` in the front).
However, if we don't to that I think should sub `nearest` with `near` (i.e., `nearTypeName` and `otherNearTypeName`) since technically one of the types might not be "nearest".